README
The goal of this project is to develop a series of terminal-based games that I will eventually
combine into a single program with an interface where the user selects which game they would like
to play. The MakeFiles of these projects are based off those of my COM S 327 class at Iowa State
University.

Hangman/CHANGELOG
	Apr 25: Started assignment 2 idea brain storming, decided on hangman and found word list
	online
	Apr 26: created logic for storing words and choosing a word at random; developed blank
	space generation; started work for ncurses
	Apr 29: implemented guessing mechanic without output; created input validation
	Apr 30: finalized word guessing mechanics; added edge cases and exit mechanic
	May 1: created hangman in nurses and logic to display; displayed letters used list;
	commented; wrote README; submitted assignment 2
	May 5: added color; added score; fixed formatting; resubmitted assignment 2


Wordle/wordle.cpp
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <sys/time.h>
#include <assert.h>
#include <unistd.h>
#include <ncurses.h>
#include <ctype.h>

#include <string>
#include <iostream>
#include <fstream>
#include <vector>
#include <cstring>
#include <algorithm>

// stores the potential words
std::vector<std::string> wordList;

int countLines(std::string s) {
  // counts the number of lines in the file
  int numLines = 0;
  std::fstream in (s);
  std::string param;
  while (std::getline(in, param)) {
    numLines++;
  }
  in.close();
  return numLines;
}

void readFile() {
  // reads in the words in the file separated by a newline
  std::string s("words.txt");
  std::fstream input (s);
  int numLines = countLines(s);
  int line = 0;
  while(line < numLines) {
    std::string word;
    getline(input, word);
    wordList.push_back(word);
    line++;
  }
  input.close();
}

// initializes the terminal for ncurses
void io_init_terminal(void) {
  initscr();
  raw();
  noecho();
  curs_set(0);
  keypad(stdscr, TRUE);
  start_color();
  // creates the color pairs
  init_pair(COLOR_RED, COLOR_RED, COLOR_BLACK);
  init_pair(COLOR_GREEN, COLOR_GREEN, COLOR_BLACK);
  init_pair(COLOR_YELLOW, COLOR_YELLOW, COLOR_BLACK);
  init_pair(COLOR_BLUE, COLOR_BLUE, COLOR_BLACK);
  init_pair(COLOR_MAGENTA, COLOR_MAGENTA, COLOR_BLACK);
  init_pair(COLOR_CYAN, COLOR_CYAN, COLOR_BLACK);
  init_pair(COLOR_WHITE, COLOR_WHITE, COLOR_BLACK);
}

void gameLoop() {
  int exit = 1;
  int guesses = 0;
  int score = 0;
  int line = 3;
  // chooses a random word
  std::string s("words.txt");
  int random = rand() % countLines(s) + 1;
  std::string currentWord = wordList.at(random);
  while (exit || guesses < 5) {
    // prompts the user
    mvprintw(0, 0, "Welcome to Wordle! Enter a word to start playing");
    std::string points = "Score: " + std::to_string(score);
    char* scoreCount = const_cast<char*>(points.c_str());
    mvprintw(1, 0, scoreCount);
    // prints the attempt number
    std::string attempt = "Guess " + std::to_string(guesses + 1) + ":";
    char* attemptCount = const_cast<char*>(attempt.c_str());
    mvprintw(line++, 0, attemptCount);
    // gets the user's guess
    std::string guess;
    int letter = getch();
    int i = 0;
    while (guess.length() < currentWord.length() - 1) {
      if (letter >= 97 && letter <= 122) {
	guess.push_back(letter);
	mvaddch(line, i, guess.at(i));
	i++;
      } else {
	// alerts the user of invalid input
	attron(COLOR_PAIR(COLOR_RED));
	line++;
	mvprintw(line++, 0, "Error, please only enter lower case letters");
	attroff(COLOR_PAIR(COLOR_RED));
      }
      // checks for final letter entered
      if (guess.length() < currentWord.length() - 1) {
	letter = getch();
      }
    }
    // checks if a valid word was entered
    int found = 0;
    for (i = 0; i < (int) wordList.size(); i++) {
      // gets the formatted word
      std::string temp;
      std::string current = wordList.at(i);
      int j;
      for (j = 0; j < (int) current.length() - 1; j++) {
	temp += current.at(j);
      }
      if (temp == guess) {
	found = 1;
      }  
    }
    // checks if the word was found
    if (found == 0) {
      line++;
      mvprintw(line++, 0, "Error, please enter a valid word");
      guess = "";
      continue;
    }
    // increments the guesses
    guesses++;
    int correct = 0;
    // checks which letters were correct or ine wrong place
    for (i = 0; i < (int) guess.length(); i++) {
      if (guess.at(i) == currentWord.at(i)) {
	attron(COLOR_PAIR(COLOR_GREEN));
	mvaddch(line, i, guess.at(i));
	attroff(COLOR_PAIR(COLOR_GREEN));
	correct++;
      } else if (currentWord.std::string::find(guess.at(i)) != std::string::npos) {
	attron(COLOR_PAIR(COLOR_YELLOW));
	mvaddch(line, i, guess.at(i));
	attroff(COLOR_PAIR(COLOR_YELLOW));
      } else {
	mvaddch(line, i, guess.at(i));
      }
    }
    line++;
    // checks of the user guessed the word correctly
    if (correct == 5) {
      score++;
      attron(COLOR_PAIR(COLOR_GREEN));
      mvprintw(line++, 0, "Correct!");
      attroff(COLOR_PAIR(COLOR_GREEN));
      line++;
      // prompts the user to play again
      mvprintw(line++, 0, "Press any key to play again or 'Q' to quit");
      int key = getch();
      if (key == 'Q') {
	// exits the game
	exit = 0;
	break;
      } else {
	// chooses a new word and resents the terminal
	random = rand() % countLines(s) + 1;
	currentWord = wordList.at(random);
	guesses = 0;
	line = 3;
	clear();
      }
    } else if (guesses >= 6) {
      // alerts the user of a loss
      std::string fail = "You ran out of guesses, the word was " + currentWord;
      char* failureMessage = const_cast<char*>(fail.c_str());
      attron(COLOR_PAIR(COLOR_RED));
      mvprintw(line++, 0, failureMessage);
      attroff(COLOR_PAIR(COLOR_RED));
      line++;
      // prompts the user to play again
      mvprintw(line++, 0, "Press any key to play again or 'Q' to quit");
      int key = getch();
      if (key == 'Q') {
	// exits the game
        exit = 0;
	break;
      } else {
	// chooses a new word and resets the terminal
	random = rand() % countLines(s) + 1;
	currentWord = wordList.at(random);
	guesses = 0;
	line = 3;
	clear();
      }
    }
  }
}

int main(int argc, char *argv[]) {
  io_init_terminal();
  srand(time(NULL));
  // reads file input
  readFile();
  // starts the game
  gameLoop();
  endwin();
  return 0;
}


TicTacToe/tictactoe.cpp
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <sys/time.h>
#include <assert.h>
#include <unistd.h>
#include <ncurses.h>
#include <ctype.h>

#include <string>
#include <iostream>
#include <fstream>
#include <vector>
#include <cstring>
#include <algorithm>

// stores the board
char board[3][3];
// stores the current line of the terminal
int line = 0;
// stores each side's score
int xScore = 0;
int oScore = 0;

void initBoard() {
  // fills the board with placeholder integers
  int number = 1;
  int i;
  for (i = 0; i < 3; i++) {
    int j;
    for (j = 0; j < 3; j++) {
      board[i][j] = number;
      number++;
    }
  }
}

void io_init_terminal(void) {
  initscr();
  raw();
  noecho();
  curs_set(0);
  keypad(stdscr, TRUE);
  start_color();
}

void printBoard() {
  line++;
  // formats the output
  std::string line1 = " " + std::to_string(board[0][0]) + " | " + std::to_string(board[0][1])
    + " | " + std::to_string(board[0][2]) + " ";
  int i;
  // replaces ASCII values with correct character
  for (i = 0; i < (int) line1.length() - 1; i++) {
    if (line1[i] == '8' && line1[i + 1] == '8') {
      line1[i] = 'X';
      line1.erase(i + 1, 1);
    } else if (line1[i] == '7' && line1[i + 1] == '9') {
      line1[i] = 'O';
      line1.erase(i + 1, 1);
    }
  }
  char* row1  = const_cast<char*>(line1.c_str());
  std::string line2 = " " + std::to_string(board[1][0]) + " | " + std::to_string(board[1][1])
    + " | " + std::to_string(board[1][2]) + " ";
  // replaces ASCII values with correct character
  for (i = 0; i < (int) line2.length() - 1; i++) {
    if (line2[i] == '8' && line2[i + 1] == '8') {
      line2[i] = 'X';
      line2.erase(i + 1, 1);
    } else if (line2[i] == '7' && line2[i + 1] == '9') {
      line2[i] = 'O';
      line2.erase(i + 1, 1);
    }
  }
  char* row2  = const_cast<char*>(line2.c_str());
  std::string line3 = " " + std::to_string(board[2][0]) + " | " + std::to_string(board[2][1])
    + " | " + std::to_string(board[2][2]) + " ";
  // replaces ASCII values with correct character
  for (i = 0; i < (int) line3.length() - 1; i++) {
    if (line3[i] == '8' && line3[i + 1] == '8') {
      line3[i] = 'X';
      line3.erase(i + 1, 1);
    } else if (line3[i] == '7' && line3[i + 1] == '9') {
      line3[i] = 'O';
      line3.erase(i + 1, 1);
    }
  }
  char* row3  = const_cast<char*>(line3.c_str());
  // prints the output
  mvprintw(line++, 0, "   |   |   ");
  mvprintw(line++, 0, row1);
  mvprintw(line++, 0, "   |   |   ");
  mvprintw(line++, 0, "---+---+--- ");
  mvprintw(line++, 0, "   |   |   ");
  mvprintw(line++, 0, row2);
  mvprintw(line++, 0, "   |   |   ");
  mvprintw(line++, 0, "---+---+---");
  mvprintw(line++, 0, "   |   |   ");
  mvprintw(line++, 0, row3);
  mvprintw(line++, 0, "   |   |   ");
  line++;
}

char checkWin() {
  if (board[0][0] == board[0][1] && board[0][0] == board[0][2]) {
    // checks first row
    return board[0][0];
  } else if (board[1][0] == board[1][1] && board[1][0] == board[1][2]) {
    // checks second row
    return board[1][0];
  } else if (board[2][0] == board[2][1] && board[2][0] == board[2][2]) {
    // checks third row
    return board[2][0];
  } else if (board[0][0] == board[1][0] && board[0][0] == board[2][0]) {
    // checks first column
    return board[0][0];
  } else if (board[0][1] == board[1][1] && board[0][1] == board[2][1]) {
    // chekcs second column
    return board[0][1];
  } else if (board[0][2] == board[1][2] && board[0][2] == board[2][2]) {
    // checks third column
    return board[0][2];
  } else if (board[0][0] == board[1][1] && board[0][0] == board[2][2]) {
    // checks top left to bottom right
    return board[0][0];
  } else if (board[0][2] == board[1][1] && board[0][2] == board[2][0]) {
    // checks top right to bottom left
    return board[0][2];
  } else if ((board[0][0] == 'X' || board[0][0] == 'O')
	     && (board[0][1] == 'X' || board[0][1] == 'O')
	     && (board[0][2] == 'X' || board[0][2] == 'O')
	     && (board[1][0] == 'X' || board[1][0] == 'O')
	     && (board[1][1] == 'X' || board[1][1] == 'O')
	     && (board[1][2] == 'X' || board[1][2] == 'O')
	     && (board[2][0] == 'X' || board[2][0] == 'O')
	     && (board[2][1] == 'X' || board[2][1] == 'O')
	     && (board[2][2] == 'X' || board[2][2] == 'O')) {
    // checks if catsgame
    return 'C';
  } else {
    // returns if game is stil in progress
    return 'P';
  }
}

void gameLoop() {
  int end = 1;
  int i = 1;
  while (end) {
    // prints header
    mvprintw(line++, 0, "Welocme to Tic Tac Toe! Enter the location you would like to play at");
    mvprintw(line++, 0, "Press 'Q' at any time to exit");
    line++;
    std::string x = "X's wins: " + std::to_string(xScore);
    char* xWins = const_cast<char*>(x.c_str());
    mvprintw(line++, 0, xWins);
    std::string o = "O's wins: " + std::to_string(oScore);
    char* oWins = const_cast<char*>(o.c_str());
    mvprintw(line++, 0, oWins);
    line++;
    printBoard();
    std::string turn;
    char symbol;
    // checks who's turn it is
    if (i % 2 == 1) {
      turn = "X's Turn:";
      symbol = 'X';
    } else {
      turn = "O's Turn:";
      symbol = 'O';
    }
    // prints the turn
    char* prompt = const_cast<char*>(turn.c_str());
    mvprintw(line++, 0, prompt);
    int input = getch();
    int loop = 1;
    // gets user input and assigns it if applicable
    do {
      switch (input) {
      case 'Q':
	end = 0;
	loop = 0;
	break;
      case 49:
	if (board[0][0] != 'X' && board[0][0] != 'O') {
	  board[0][0] = symbol;
	  loop = 0;
	} else {
	  mvprintw(line++, 0, "Error, that position is already taken. Try again:");
	  input = getch();
	}
	break;
      case 50:
	if (board[0][1] != 'X' && board[0][1] != 'O') {
	  board[0][1] = symbol;
	  loop = 0;
	} else {
	  mvprintw(line++, 0, "Error, that position is already taken. Try again:");
	  input = getch();
	}
	break;
      case 51:
	if (board[0][2] != 'X' && board[0][2] != 'O') {
	  board[0][2] = symbol;
	  loop = 0;
	} else {
	  mvprintw(line++, 0, "Error, that position is already taken. Try again:");
	  input = getch();
	}
	break;
      case 52:
	if (board[1][0] != 'X' && board[1][0] != 'O') {
	  board[1][0] = symbol;
	  loop = 0;
	} else {
	  mvprintw(line++, 0, "Error, that position is already taken. Try again:");
	  input = getch();
	}
	break;
      case 53:
	if (board[1][1] != 'X' && board[1][1] != 'O') {
	  board[1][1] = symbol;
	  loop = 0;
	} else {
	  mvprintw(line++, 0, "Error, that position is already taken. Try again:");
	  input = getch();
	}
	break;
      case 54:
	if (board[1][2] != 'X' && board[1][2] != 'O') {
	  board[1][2] = symbol;
	  loop = 0;
	} else {
	  mvprintw(line++, 0, "Error, that position is already taken. Try again:");
	  input = getch();
	}
	break;
      case 55:
	if (board[2][0] != 'X' && board[2][0] != 'O') {
	  board[2][0] = symbol;
	  loop = 0;
	} else {
	  mvprintw(line++, 0, "Error, that position is already taken. Try again:");
	  input = getch();
	}
	break;
      case 56:
	if (board[2][1] != 'X' && board[2][1] != 'O') {
	  board[2][1] = symbol;
	  loop = 0;
	} else {
	  mvprintw(line++, 0, "Error, that position is already taken. Try again:");
	  input = getch();
	}
	break;
      case 57:
	if (board[2][2] != 'X' && board[2][2] != 'O') {
	  board[2][2] = symbol;
	  loop = 0;
	} else {
	  mvprintw(line++, 0, "Error, that position is already taken. Try again:");
	  input = getch();
	}
	break;
      default:
	mvprintw(line++, 0, "Error, please only enter numbers 1-9. Try again:");
	input = getch();
      }
    } while (loop);
    // checks if the game is over
    char win = checkWin();
    if (win == 'C') {
      clear();
      line = 0;
      // prints the header
      mvprintw(line++, 0, "Welocme to Tic Tac Toe! Enter the location you would like to play at");
      mvprintw(line++, 0, "Press 'Q' at any time to exit");
      line++;
      mvprintw(line++, 0, xWins);
      mvprintw(line++, 0, oWins);
      line++;
      printBoard();
      // prints the result
      mvprintw(line++, 0, "Game over, it was a tie!");
      mvprintw(line++, 0, "Press any key to play again, 'Q' to quit");
      // prompts the user to play again
      int key = getch();
      if (key == 'Q') {
	end = 0;
      } else {
	initBoard();
      }
    } else if (win == 'P') {
      // moves on with the game
      i++;
    } else {
      clear();
      line = 0;
      // prints the header
      mvprintw(line++, 0, "Welocme to Tic Tac Toe! Enter the location you would like to play at");
      mvprintw(line++, 0, "Press 'Q' at any time to exit");
      line++;
      mvprintw(line++, 0, xWins);
      mvprintw(line++, 0, oWins);
      line++;
      printBoard();
      // prints the result
      std::string winStr = "Game over, " + std::to_string(win) + "'s win!";
      // replaces ASCII values with correct character
      for (i = 0; i < (int) winStr.length() - 1; i++) {
	if (winStr[i] == '8' && winStr[i + 1] == '8') {
	  winStr[i] = 'X';
	  xScore++;
	  winStr.erase(i + 1, 1);
	} else if (winStr[i] == '7' && winStr[i + 1] == '9') {
	  winStr[i] = 'O';
	  oScore++;
	  winStr.erase(i + 1, 1);
	}
      }
      char* winMessage = const_cast<char*>(winStr.c_str());
      mvprintw(line++, 0, winMessage);
      mvprintw(line++, 0, "Press any key to play again, 'Q' to quit");
      // prompts the user to play again
      int key = getch();
      if (key == 'Q') {
	end = 0;
      } else {
	initBoard();
      }
    }
    // resets the terminal
    clear();
    line = 0;
  }
}

int main(int argc, char *argv[]) {
  io_init_terminal();
  initBoard();
  // starts the game
  gameLoop();
  endwin();
  return 0;
}


Hangman/hangman.cpp
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <limits.h>
#include <sys/time.h>
#include <assert.h>
#include <unistd.h>
#include <ncurses.h>
#include <ctype.h>

#include <string>
#include <iostream>
#include <fstream>
#include <vector>
#include <cstring>

// stroes the potential words
std::vector<std::string> wordList;
int score = 0;

int countLines(std::string s) {
  // counts the number of lines in the file
  int numLines = 0;
  std::fstream in (s);
  std::string param;
  while (std::getline(in, param)) {
    numLines++;
  }
  in.close();
  return numLines;
}

void readFile() {
  // reads in the words in the file separated by a newline
  std::string s("words.txt");
  std::fstream input (s);
  int numLines = countLines(s);
  int line = 0;
  while(line < numLines) {
    std::string word;
    getline(input, word);
    wordList.push_back(word);
    line++;
  }
  input.close();
}

// initializes the terminal for ncurses
void io_init_terminal(void)
{
  initscr();
  raw();
  noecho();
  curs_set(0);
  keypad(stdscr, TRUE);
  start_color();
  init_pair(COLOR_RED, COLOR_RED, COLOR_BLACK);
  init_pair(COLOR_GREEN, COLOR_GREEN, COLOR_BLACK);
  init_pair(COLOR_YELLOW, COLOR_YELLOW, COLOR_BLACK);
  init_pair(COLOR_BLUE, COLOR_BLUE, COLOR_BLACK);
  init_pair(COLOR_MAGENTA, COLOR_MAGENTA, COLOR_BLACK);
  init_pair(COLOR_CYAN, COLOR_CYAN, COLOR_BLACK);
  init_pair(COLOR_WHITE, COLOR_WHITE, COLOR_BLACK);
}

// prints the hangman based on the amount of guesses used
void printHangman(int numGuesses) {
  int i;
  // clears the terminal
  for (i = 0; i < 14; i++) {
     mvprintw(i, 0, "                                                     ");
  }
  switch(numGuesses) {
  case 0:
    mvprintw(0, 0, " ________");
    mvprintw(1, 0, " |      |");
    mvprintw(2, 0, " |      |");
    mvprintw(3, 0, " |");
    mvprintw(4, 0, " |");
    mvprintw(5, 0, " |");
    mvprintw(6, 0, " |");
    mvprintw(7, 0, " |");
    mvprintw(8, 0, " |");
    mvprintw(9, 0, " |");
    mvprintw(10, 0, " |");
    mvprintw(11, 0, " |");
    mvprintw(12, 0, " |");
    mvprintw(13, 0, "___");
    break;
  case 1:
     mvprintw(0, 0, " ________");
     mvprintw(1, 0, " |      |");
     mvprintw(2, 0, " |      |");
     mvprintw(3, 0, " |    /---\\");
     mvprintw(4, 0, " |    |   |");
     mvprintw(5, 0, " |    \\---/");
     mvprintw(6, 0, " |");
     mvprintw(7, 0, " |");
     mvprintw(8, 0, " |");
     mvprintw(9, 0, " |");
     mvprintw(10, 0, " |");
     mvprintw(11, 0, " |");
     mvprintw(12, 0, " |");
     mvprintw(13, 0, "___");
    break;
  case 2:
    mvprintw(0, 0, " ________");
    mvprintw(1, 0, " |      |");
    mvprintw(2, 0, " |      |");
    mvprintw(3, 0, " |    /---\\");
    mvprintw(4, 0, " |    |   |");
    mvprintw(5, 0, " |    \\---/");
    mvprintw(6, 0, " |      |");
    mvprintw(7, 0, " |      |");
    mvprintw(8, 0, " |      |");
    mvprintw(9, 0, " |      |");
    mvprintw(10, 0, " |");
    mvprintw(11, 0, " |");
    mvprintw(12, 0, " |");
    mvprintw(13, 0, "___");
    break;
  case 3:
    mvprintw(0, 0, " ________");
    mvprintw(1, 0, " |      |");
    mvprintw(2, 0, " |      |");
    mvprintw(3, 0, " |    /---\\");
    mvprintw(4, 0, " |    |   |");
    mvprintw(5, 0, " |    \\---/");
    mvprintw(6, 0, " |      |");
    mvprintw(7, 0, " |   ___|");
    mvprintw(8, 0, " |      |");
    mvprintw(9, 0, " |      |");
    mvprintw(10, 0, " |");
    mvprintw(11, 0, " |");
    mvprintw(12, 0, " |");
    mvprintw(13, 0, "___");
    break;
  case 4:
    mvprintw(0, 0, " ________");
    mvprintw(1, 0, " |      |");
    mvprintw(2, 0, " |      |");
    mvprintw(3, 0, " |    /---\\");
    mvprintw(4, 0, " |    |   |");
    mvprintw(5, 0, " |    \\---/");
    mvprintw(6, 0, " |      |");
    mvprintw(7, 0, " |   ___|___");
    mvprintw(8, 0, " |      |");
    mvprintw(9, 0, " |      |");
    mvprintw(10, 0, " |");
    mvprintw(11, 0, " |");
    mvprintw(12, 0, " |");
    mvprintw(13, 0, "___");
    break;
  case 5:
    mvprintw(0, 0, " ________");
    mvprintw(1, 0, " |      |");
    mvprintw(2, 0, " |      |");
    mvprintw(3, 0, " |    /---\\");
    mvprintw(4, 0, " |    |   |");
    mvprintw(5, 0, " |    \\---/");
    mvprintw(6, 0, " |      |");
    mvprintw(7, 0, " |   ___|___");
    mvprintw(8, 0, " |      |");
    mvprintw(9, 0, " |      |");
    mvprintw(10, 0, " |     /");
    mvprintw(11, 0, " |    /");
    mvprintw(12, 0, " |");
    mvprintw(13, 0, "___");
    break;
  case 6:
    mvprintw(0, 0, " ________");
    mvprintw(1, 0, " |      |");
    mvprintw(2, 0, " |      |");
    mvprintw(3, 0, " |    /---\\");
    mvprintw(4, 0, " |    |   |");
    mvprintw(5, 0, " |    \\---/");
    mvprintw(6, 0, " |      |");
    mvprintw(7, 0, " |   ___|___");
    mvprintw(8, 0, " |      |");
    mvprintw(9, 0, " |      |");
    mvprintw(10, 0, " |     / \\");
    mvprintw(11, 0, " |    /   \\");
    mvprintw(12, 0, " |");
    mvprintw(13, 0, "___");
    break;
  }
}

int gameLoop() {
  mvprintw(19, 0, "                                                     ");
  mvprintw(19, 0, "Press 'Q' to exit at any time");
  std::vector<char> lettersUsed;
  int numGuesses = 0;
  // chooses a random word from the list
  std::string s("words.txt");
  int random = rand() % countLines(s) + 1;
  std::string currentWord = wordList.at(random);
  std::string blank;
  std::string guess;
  int i = 0;
  // creates the blank spaces for the selected word
  for (i = 0; i < (int) currentWord.length(); i++) {
    blank += "_";
    guess += "_";
    if (i < (int) currentWord.length() - 1) {
      blank += " ";
    }
  }
  // converts the blank for ncurses
  char* curseBlank = const_cast<char*>(blank.c_str());
  mvprintw(15, 0, curseBlank);
  int correct = 0;
  // prints the letters used
  while ((int) lettersUsed.size() < 26 && numGuesses < 6 && correct == 0) {
    std::string used = "Letters used: ";
    for (i = 0; i < (int) lettersUsed.size(); i++) {
      used += lettersUsed.at(i);
      if (i < (int) lettersUsed.size() - 1) {
	used += " ";
      }
    }
    // converts the letters used for ncurses
    char* curseUsed = const_cast<char*>(used.c_str());
    mvprintw(20, 0, "                                                     ");
    attron(COLOR_PAIR(COLOR_RED));
    mvprintw(20, 0, curseUsed);
    attroff(COLOR_PAIR(COLOR_RED));
    // prints the score
    std::string scoreOut = "Score: " + std::to_string(score);
    char* curseScore = const_cast<char*>(scoreOut.c_str());
    mvprintw(21, 0, "                                                     ");
    attron(COLOR_PAIR(COLOR_CYAN));
    mvprintw(21, 0, curseScore);
    attroff(COLOR_PAIR(COLOR_CYAN));
    int key;
    int letter = '!';
    // gets user input for their guessed letter
    while (letter == '!') {
      key = getch();
      // exits the game
      if (key == 'Q') {
	 mvprintw(15, 0, "                                                     ");
	 mvprintw(18, 0, "                                                     ");
	 mvprintw(19, 0, "                                                     ");
	 mvprintw(17, 0, "                                                     ");
	 mvprintw(20, 0, "                                                     ");
	 mvprintw(21, 0, "                                                     ");
	 attron(COLOR_PAIR(COLOR_YELLOW));
	 mvprintw(17, 0, "Game exited by user, press 'Q' to close");
	 attroff(COLOR_PAIR(COLOR_YELLOW));
	 return -1;
      }
      // checks for valid key
      if (key >= 97 && key <= 122) {
	int used = 0;
	// checks if the letter has been used
	for (i = 0; i < (int) lettersUsed.size(); i++) {
	  if (lettersUsed.at(i) == key) {
	    used = 1;
	  }
	}
	mvprintw(17, 0, "                                                       ");
	// alerts the user they have selected a duplicate letter
	if (used == 1) {
	  mvprintw(17, 0, "                                                     ");
	  attron(COLOR_PAIR(COLOR_YELLOW));
	  mvprintw(17, 0, "You already guessed that letter, try again");
	  attroff(COLOR_PAIR(COLOR_YELLOW));
	} else {
	  letter = key;
	  lettersUsed.push_back(letter);
	  int wrong = 0;
	  // checks if the user uses a guess incorrectly
	  for (i = 0; i < (int) currentWord.length(); i++) {
	    if (currentWord[i] == letter) {
	      wrong = 1;
	    }
	  }
	  if (wrong == 0) {
	    numGuesses++;
	    printHangman(numGuesses);
	  }
	}
      } else {
	// alerts the user of invalid input
	mvprintw(17, 0, "                                                       ");
	attron(COLOR_PAIR(COLOR_YELLOW));
	mvprintw(17, 0, "Please only enter letters a-z lower case");
	attroff(COLOR_PAIR(COLOR_YELLOW));
      }
    }
    // checks where the user's guess is in the word
    for (i = 0; i < (int) lettersUsed.size(); i++) {
      int j;
      for (j = 0; j < (int) currentWord.length(); j++) {
	if (currentWord[j] == lettersUsed.at(i)) {
	  guess[j] = lettersUsed.at(i);
	}
      }
    }
    // updates the blank with the guessed letters
    std::string guessOut;
    for (i = 0; i < (int) guess.length(); i++) {
      guessOut += guess[i];
      if (i < (int) guess.length() - 1) {
	guessOut += " ";
      }
    }
    int numLeft = 0;
    for (i = 0; i < (int) guess.length(); i++) {
      if (guess[i] == '_') {
	numLeft++;
      }
    }
    // checks if the user guessed the word correctly
    if (numLeft == 0) {
      correct = 1;
      mvprintw(17, 0, "                                                     ");
      attron(COLOR_PAIR(COLOR_GREEN));
      mvprintw(17, 0, "You have guessed the word correctly");
      attroff(COLOR_PAIR(COLOR_GREEN));
      // incremets the score
      score++;
    }
    // converts the user's guesses for ncurses
    char* curseGuess = const_cast<char*>(guessOut.c_str());
    mvprintw(15, 0, "                                                     ");
    mvprintw(15, 0, curseGuess);
  }
  // checks if the user completed the hangman
  if (numGuesses >= 6) {
     mvprintw(17, 0, "                                                     ");
     attron(COLOR_PAIR(COLOR_YELLOW));
     mvprintw(17, 0, "You have reached the maximum number of guesses");
     attroff(COLOR_PAIR(COLOR_YELLOW));
  }    
  std::string answer = "The word was " + currentWord;
  // converts the answer for ncurses
  char* curseCurrent = const_cast<char*>(answer.c_str());
  mvprintw(18, 0, "                                                     ");
  mvprintw(18, 0, curseCurrent);
  mvprintw(19, 0, "                                                     ");
  mvprintw(19, 0, "Press any key to play again or 'Q' to exit");
  return 0;
}

int main(int argc, char *argv[]) {
  io_init_terminal();
  srand(time(NULL));
  readFile();
  int loop = 0;
  // loops until the user presses Q
  while (loop == 0) {
    mvprintw(15, 0, "                                                     ");
    mvprintw(17, 0, "                                                     ");
    mvprintw(18, 0, "                                                     ");
    mvprintw(19, 0, "                                                     ");
    printHangman(0);
    loop = gameLoop();
    getch();
  }
  // ends the ncurses terminal
  endwin();
  return 0;
}
